### ISSUE TITLE
`pl.read_database()` with `iter_batches=True` no longer loads batches incrementally

### ISSUE THREAD

**Issue by Kigstn** [pola-rs/polars#15469]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
# please change
table_name = "TABLE_NAME"
conn = "SQL_ALCH_CONN"

with pl.Config(verbose=True):  
    dfs = pl.read_database(query=f"SELECT * FROM {table_name}", connection=conn, iter_batches=True, batch_size=100_000)
    f_name = f"output/{table_name}.csv"
    with open(f_name, mode="ab") as f:
        for i, df in enumerate(dfs):
            df.write_csv(f, separator="|", include_header=False)
            print(f"Written {table_name} - {i}")
```

### Log output

```shell
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
Cell In[6], line 9
      5 with pl.Config(verbose=True):  
      6 # please change
      7     table_name = "DMD_CONTRACT_PART_LINK "
----> 9     dfs = pl.read_database(query=f"SELECT * FROM {table_name}", connection=conn, iter_batches=True, batch_size=100_000)
     10     f_name = f"test/{table_name}.csv"
     11     with open(f_name, mode="ab") as f:

File C:\Program Files\Python311\Lib\site-packages\polars\io\database\functions.py:267, in read_database(query, connection, iter_batches, batch_size, schema_overrides, infer_schema_length, execute_options, **kwargs)
    262 # return frame from arbitrary connections using the executor abstraction
    263 with ConnectionExecutor(connection) as cx:
    264     return cx.execute(
    265         query=query,
    266         options=execute_options,
--> 267     ).to_polars(
    268         batch_size=batch_size,
    269         iter_batches=iter_batches,
    270         schema_overrides=schema_overrides,
    271         infer_schema_length=infer_schema_length,
    272     )

File C:\Program Files\Python311\Lib\site-packages\polars\io\database\_executor.py:508, in ConnectionExecutor.to_polars(self, iter_batches, batch_size, schema_overrides, infer_schema_length)
    502     raise ValueError(msg)
    504 for frame_init in (
    505     self._from_arrow,  # init from arrow-native data (where support exists)
    506     self._from_rows,  # row-wise fallback (sqlalchemy, dbapi2, pyodbc, etc)
    507 ):
--> 508     frame = frame_init(
    509         batch_size=batch_size,
    510         iter_batches=iter_batches,
    511         schema_overrides=schema_overrides,
    512         infer_schema_length=infer_schema_length,
    513     )
    514     if frame is not None:
    515         return frame

File C:\Program Files\Python311\Lib\site-packages\polars\io\database\_executor.py:279, in ConnectionExecutor._from_rows(self, batch_size, iter_batches, schema_overrides, infer_schema_length)
    265     schema_overrides = self._inject_type_overrides(
    266         description=cursor_desc,
    267         schema_overrides=(schema_overrides or {}),
    268     )
    269     result_columns = list(cursor_desc)
    270     frames = (
    271         DataFrame(
    272             data=rows,
    273             schema=result_columns or None,
    274             schema_overrides=schema_overrides,
    275             infer_schema_length=infer_schema_length,
    276             orient="row",
    277         )
    278         for rows in (
--> 279             list(self._fetchmany_rows(self.result, batch_size))
    280             if iter_batches
    281             else [self._fetchall_rows(self.result)]  # type: ignore[list-item]
    282         )
    283     )
    284     return frames if iter_batches else next(frames)  # type: ignore[arg-type]
    285 return None

File C:\Program Files\Python311\Lib\site-packages\polars\io\database\_executor.py:185, in ConnectionExecutor._fetchmany_rows(self, result, batch_size)
    183 """Fetch row data incrementally, yielding over the complete result set."""
    184 while True:
--> 185     rows = result.fetchmany(batch_size)
    186     if not rows:
    187         break

File C:\Program Files\Python311\Lib\site-packages\sqlalchemy\engine\result.py:1371, in Result.fetchmany(self, size)
   1352 def fetchmany(self, size: Optional[int] = None) -> Sequence[Row[_TP]]:
   1353     """Fetch many rows.
   1354 
   1355     When all rows are exhausted, returns an empty list.
   (...)
   1368 
   1369     """
-> 1371     return self._manyrow_getter(self, size)

File C:\Program Files\Python311\Lib\site-packages\sqlalchemy\engine\result.py:724, in ResultInternal._manyrow_getter.<locals>.manyrows(self, num)
    722 rows: List[_InterimRowType[Any]] = self._fetchmany_impl(num)
    723 if make_row:
--> 724     rows = [make_row(row) for row in rows]
    725 if post_creational_filter:
    726     rows = [post_creational_filter(row) for row in rows]

File C:\Program Files\Python311\Lib\site-packages\sqlalchemy\engine\result.py:724, in <listcomp>(.0)
    722 rows: List[_InterimRowType[Any]] = self._fetchmany_impl(num)
    723 if make_row:
--> 724     rows = [make_row(row) for row in rows]
    725 if post_creational_filter:
    726     rows = [post_creational_filter(row) for row in rows]

MemoryError:
```

### Issue description

On older versions (tested with `0.20.6`) this code would load the batch into memory and write it. Then load the next batch while deleting the old batch from memory.
This was needed in my use case, because the data I am working with is really big with a couple 100m rows and I have 16GB of RAM on this machine.

With the latest version (`0.20.18`), this code tries to load all the batches into memory before writing it. When dealing with large tables, this results in a `MemoryError`. 
At least this is what I assume, considering the task manger RAM usage:
![grafik](https://github.com/pola-rs/polars/assets/38250010/f69bf07c-cd48-403e-a00e-5208436875e4)

### Expected behavior

Only load the batch into memory, not the entire result.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.18
Index type:           UInt32
Platform:             Windows-10-10.0.22000-SP0
Python:               3.11.3 (tags/v3.11.3:f3909b8, Apr  4 2023, 23:49:59) [MSC v.1934 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.3
nest_asyncio:         1.6.0
numpy:                1.23.4
openpyxl:             3.1.2
pandas:               2.2.0
pyarrow:              12.0.1
pydantic:             1.10.12
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.19
xlsx2csv:             0.8.2
xlsxwriter:           <not installed>
```

</details>

**Comment by alexander-beedie** [comment]
Can you confirm which oracle driver you're using (cx_Oracle?), along with any connection settings you're passing to sqlalchemy?

**Comment by alexander-beedie** [comment]
Ahh! I've spotted the problem... will make a PR to fix it shortly ✌️

**Comment by alexander-beedie** [comment]
@Kigstn: fix is in the new `0.20.19` release 👌
